### ISSUE TITLE
Changing one item in O2M corrupting data in other items in the same O2M relation if you click on them before saving at top level

### ISSUE THREAD

**Issue by jpetryk** [directus/directus#5752]
@MiniDigger thanks for the suggestion to check the display template. In my case, it already has one. I can also verify that when I change one item, save, go into the other item (where it displays the change to the field from the first one), it comes back from the API properly. It's like it believes I'm making the change from the first one in all of them though.

To demonstrate, the attached screenshot shows the API call for the item with ID 22 being correct. Yet the interface is displaying the change I made to ID 21. It actually doesn't matter if, after making the change to the first one, I save it or just click the X. The damage is done. All will be displaying the tentative change if I go in them. Again, if I don't go into any others and save it all the way through to the top level, it will only make the expected change. The oddities occur only after I click on any others prior to saving.

The workaround is having to make changes one at a time, being careful to not click anything other than the one I am changing, saving at the top level each time. Obviously not ideal.

This is a different issue because setting the display template doesn't change anything, and I _think_ my issue actually affects O2M (and perhaps not others). As a result, I'll peel this off into a new issue.

![Screenshot 2021-05-18 22 32 04](https://user-images.githubusercontent.com/5734619/118757699-d4027480-b82a-11eb-8a9d-749e4205b860.png)

_Originally posted by @jpetryk in https://github.com/directus/directus/issues/5681#issuecomment-843742689_

**Comment by jpetryk** [comment]
Here's the link to my original video in the other issue in case it helps: https://www.loom.com/share/86c6a9ee913b4361ae25775989532248

**Comment by maximilian-github** [comment]
I ran into similar issues today when developing a custom o2m interface. I think in my case the problem was that newly added items do not have a primaryKey and instead the plus sign is being used to identify them. So `editsAtStart` will always hold the values of the first item in the list. 
```
editsAtStart.value = edits || { [pkField]: item[pkField] || {} };
currentlyEditing.value = hasPrimaryKey ? item[pkField] : '+';
```

This also leads to another issue: If (newly added) related items have the exact same values for all fields, then deleting one of them will also remove all other items. 

I 'solved' this by using the index of `v-list-item` in a similar way (but with that it all gets kind of weird): 
```
currentlyEditingIndex.value = index;
```

You can reproduce these issues easily just with the standard o2m interface. Also I noted that vue re-renders the entire list when an item has changed. I think that shouldn't be the case … maybe those things are related?

Here's my use case: a custom interface with 'inline editing' of o2m items. Maybe it helps to keep this in mind when refactoring the o2m interface?

<img width="629" alt="Screenshot 2021-05-23 at 13 59 40" src="https://user-images.githubusercontent.com/83605655/119274843-3236ab00-bbcf-11eb-80ab-371ddea3f8fb.png">


//

Talking about the o2m interface: Could you take a look at the interface options (in settings/data-model)? I think something is broken:
```
[Vue warn]: Error in render: "TypeError: Cannot read property 'includes' of undefined"
found in
<SetupInterface> at src/modules/settings/routes/data-model/field-detail/components/interface.vue
```

**Comment by benosman** [comment]
I am also experiencing this issue, and the behaviour is very much the same as @jpetryk shows in the video with O2M fields.

I can't add much else here as it seems well covered.

**Comment by stefanvs** [comment]
@Nitwel I've provided a minimalistic db dump to get this issue reproduced.
[directus-o2m-issue---FULL-EXPORT---2021-09-09-13-02.sql.zip](https://github.com/directus/directus/files/7135996/directus-o2m-issue---FULL-EXPORT---2021-09-09-13-02.sql.zip)

And here a screenrecording of the issue:
https://user-images.githubusercontent.com/6349005/132676079-f2a8bc5e-27cc-4205-97c2-f334b89a24d4.mov

**Comment by stefanvs** [comment]
I noticed multiple things going wrong here.

```
const edits = (props.value || []).find(
  (edit: any) =>
    typeof edit === 'object' &&
    relatedPrimaryKeyField.value?.field &&
    edit[relatedPrimaryKeyField.value?.field] === item[relatedPrimaryKeyField.value?.field]
);
editsAtStart.value = edits || { [pkField]: item[pkField] || {} };
```
This piece of code results in editsAtStart to have the value of the first unsaved item, because it found the edit entry with the corresponding id (both undefined, since it's not saved yet).

Then there is also the deleting issue, which causes multiple entries to be deleted if their content is the same.
Which is caused by this:

```
if (relatedPrimKey in item === false) {
  emit(
    'input',
    props.value.filter((val) => isEqual(item, val) === false)
  );
  return;
}
```

**Comment by sousvidal** [comment]
@rijkvanzanten @Nitwel There's been quite some valuable info added to get it reproduced quickly, have you guys had any luck with this? Relational data (especially O2M) is pretty much unusable without a fix for this...

**Comment by rijkvanzanten** [comment]
@sousvidal Wanna open a PR?

**Comment by sousvidal** [comment]
@rijkvanzanten Fine.
https://github.com/directus/directus/pull/8116

It's a partial fix for most of the o2m-list issues with minimal impact (as far as I can tell, the component is a bit of a mess). I've tested it extensively with @stefanvs samples. I don't know how it will impact the issues @jpetryk has since that scenario is quite a bit more complex and there is no database export provided.

This fix does **not** get rid of the delete issue, this might require a bigger change. If this PR gets through I will be happily reluctant to take a look at that issue as well.

**Comment by kkingdude** [comment]
This issue is still happening with pre-existing children of O2M.
![CleanShot 2022-03-04 at 10 20 34](https://user-images.githubusercontent.com/1907150/156820554-770b6805-a68e-438f-a053-6525e364869f.gif)
